builder/src/core/bootstrap_dedup.rs

```diff
@@ -0,0 +1,166 @@
+    reader: RafsIoReader,
+    writer: Box<dyn RafsIoWrite>,
+    backend: BackendConfigV2,
+    encrypt: bool,
+}
+
+impl BootstrapDedup {
+    pub fn new(bootstrap_path: PathBuf, output_path: PathBuf, cfg: &Arc<ConfigV2>) -> Result<Self> {
+        let (rs, _reader) = RafsSuper::load_from_file(&bootstrap_path, cfg.clone(), false)?;
+
+        let reader = Box::new(
+            fs::OpenOptions::new()
+                .read(true)
+                .write(false)
+                .open(&bootstrap_path)?,
+        ) as RafsIoReader;
+
+        let dedup_config = cfg.get_deduplication_config()?;
+        let db_dir = dedup_config.get_work_dir()?;
+        let cas_mgr = CasMgr::new(db_dir)?;
+
+        let cache_chunks = HashMap::new();
+        let insert_chunks = vec![];
+        let insert_blobs = vec![];
+        let backend = cfg.get_backend_config().unwrap().clone();
+        let encrypt = !rs.meta.flags.contains(RafsSuperFlags::ENCRYPTION_NONE);
+        let writer = Box::new(ArtifactFileWriter(ArtifactWriter::new(
+            ArtifactStorage::SingleFile(PathBuf::from(&output_path)),
+        )?)) as Box<dyn RafsIoWrite>;
+
+        fs::copy(&bootstrap_path, &output_path)?;
+
+        Ok(BootstrapDedup {
+            cas_mgr,
+            rs,
+            cache_chunks,
+            insert_chunks,
+            insert_blobs,
+            reader,
+            writer,
+            backend,
+            encrypt,
+        })
+    }
+
+    fn is_tarfs(&self) -> bool {
+        self.rs.meta.flags.contains(RafsSuperFlags::TARTFS_MODE)
+    }
+
+    pub fn do_dedup(&mut self) -> Result<()> {
+        let conversion_type = if self.is_tarfs() {
+            ConversionType::TarToTarfs
+        } else {
+            ConversionType::DirectoryToRafs
+        };
+        let mut build_ctx = BuildContext::new(
+            "".to_string(),
+            false,
+            0,
+            self.rs.meta.get_compressor(),
+            self.rs.meta.get_digester(),
+            self.rs.meta.explicit_uidgid(),
+            WhiteoutSpec::Oci,
+            conversion_type,
+            PathBuf::from(""),
+            Default::default(),
+            None,
+            false,
+            Features::new(),
+            self.encrypt,
+        );
+
+        build_ctx.set_fs_version(RafsVersion::try_from(self.rs.meta.version).unwrap());
+
+        let is_v6 = match &build_ctx.fs_version {
+            RafsVersion::V5 => false,
+            RafsVersion::V6 => true,
+        };
+        let tree = Tree::from_bootstrap(&self.rs, &mut ())?;
+
+        let mut chunk_cache = BTreeMap::new();
+        let mut blob_mgr = BlobManager::new(digest::Algorithm::Sha256);
+        blob_mgr
+            .extend_from_blob_table(&build_ctx, self.rs.superblock.get_blob_infos())
+            .unwrap();
+
+        // Dump bootstrap
+        timing_tracer!(
+            {
+                tree.walk_bfs(true, &mut |n| {
+                    n.lock_node().dedup_chunk_for_node(
```

Comment:
`get_chunk_ofs` seems not implemented with v5. Is `dedup_chunk_for_node` able to work with v5? If not please use `is_v6` to avoid it.